src/OmniSharp.MSBuild/ProjectManager.cs

```diff
@@ -0,0 +1,469 @@
+                // No changes to make. Moving on.
+                return;
+            }
+
+            var parseOptions = new CSharpParseOptions(languageVersion);
+
+            if (preprocessorSymbolNames.Any())
+            {
+                parseOptions = parseOptions.WithPreprocessorSymbols(preprocessorSymbolNames);
+            }
+
+            if (generateXmlDocumentation)
+            {
+                parseOptions = parseOptions.WithDocumentationMode(DocumentationMode.Diagnose);
+            }
+
+            _workspace.SetParseOptions(project.Id, parseOptions);
+        }
+
+        private void UpdateProjectReferences(Project project, ImmutableArray<string> projectReferencePaths)
+        {
+            _logger.LogInformation($"Update project: {project.Name}");
+
+            var existingProjectReferences = new HashSet<ProjectReference>(project.ProjectReferences);
+            var addedProjectReferences = new HashSet<ProjectReference>();
+
+            foreach (var projectReferencePath in projectReferencePaths)
+            {
+                if (!_projectFiles.TryGetValue(projectReferencePath, out var referencedProject))
+                {
+                    if (File.Exists(projectReferencePath))
+                    {
+                        _logger.LogInformation($"Found referenced project outside root directory: {projectReferencePath}");
+
+                        // We've found a project reference that we didn't know about already, but it exists on disk.
+                        // This is likely a project that is outside of OmniSharp's TargetDirectory.
+                        referencedProject = ProjectFileInfo.CreateNoBuild(projectReferencePath, _projectLoader);
+                        AddProject(referencedProject);
+
+                        QueueProjectUpdate(projectReferencePath);
+                    }
+                }
+
+                if (referencedProject == null)
+                {
+                    _logger.LogWarning($"Unable to resolve project reference '{projectReferencePath}' for '{project.Name}'.");
+                    continue;
+                }
+
+                var projectReference = new ProjectReference(referencedProject.Id);
+
+                if (existingProjectReferences.Remove(projectReference))
+                {
+                    // This reference already exists
+                    continue;
+                }
+
+                if (!addedProjectReferences.Contains(projectReference))
+                {
+                    _workspace.AddProjectReference(project.Id, projectReference);
+                    addedProjectReferences.Add(projectReference);
+                }
+            }
+
+            foreach (var existingProjectReference in existingProjectReferences)
+            {
+                _workspace.RemoveProjectReference(project.Id, existingProjectReference);
+            }
+        }
+
+        private class MetadataReferenceComparer : IEqualityComparer<MetadataReference>
```

Comment:
I think this needs to be extracted into a reusable place - it will be useful i.e. here https://github.com/OmniSharp/omnisharp-roslyn/blob/d3930049529fe3342de8d206a18c6823dd071d96/src/OmniSharp.Script/ScriptProjectSystem.cs#L97

Reply:
Yes, good idea. I'll make it a utility in OmniSharp.Roslyn.